csharp/src/Google.Protobuf/JsonFormatter.cs

```diff
@@ -962,25 +963,17 @@ public Settings(bool formatDefaultValues, TypeRegistry typeRegistry) : this(form
         // The need for this is unfortunate, as is its unbounded size, but realistically it shouldn't cause issues.
         private static class OriginalEnumValueHelper
         {
-            // TODO: In the future we might want to use ConcurrentDictionary, at the point where all
-            // the platforms we target have it.
-            private static readonly Dictionary<System.Type, Dictionary<object, string>> dictionaries
-                = new Dictionary<System.Type, Dictionary<object, string>>();
+            private static readonly ConcurrentDictionary<System.Type, Dictionary<object, string>> dictionaries
+                = new ConcurrentDictionary<System.Type, Dictionary<object, string>>();
 
             [UnconditionalSuppressMessage("Trimming", "IL2072",
                 Justification = "The field for the value must still be present. It will be returned by reflection, will be in this collection, and its name can be resolved.")]
             internal static string GetOriginalName(object value)
             {
-                var enumType = value.GetType();
-                Dictionary<object, string> nameMapping;
-                lock (dictionaries)
+                Dictionary<object, string> nameMapping = dictionaries.GetOrAdd(value.GetType(), static t =>
```

Comment:
Any reason for not just using `static t => GetNameMapping(t)` here rather than a block body? (I *would* suggest `GetNameMapping` with a method group conversion, but I think that still creates a delegate instance per call with most compilers, annoyingly.)